keps/420-partial-admission/README.md

```diff
@@ -0,0 +1,139 @@
+
+### Non-Goals
+
+Since this is an opt-in feature, the parent job should accept the partial admission parameters provided by Kueue.
+
+Kueue will not take any measure to ensure that the parent job respects the assigned quota.
+
+## Proposal
+
+Change the way the flavor assigner work to support decrementing the pods count in order to find a better fit for the current workload.
+In case a partial fit is chosen, the jobframework reconciler should provide the admitted pod counts to the parent job before unsuspending it, in a similar fashion as the node selectors are provided. In case the job gets suspended, the original pod counts should be restored in order to allow a potential future admission with its original pod counts.
+
+### User Stories
+
+Kueue issue [420](https://github.com/kubernetes-sigs/kueue/issues/420) provides details on the initial feature details and its applicability for `batch/Job`.
+
+### Risks and Mitigations
+
+TBD: It should be decided if the partial admission should be taken into account before attempting preemption. Since preemption has its own penalties (time to evict the active workloads, potential waste of the resources used by the preempted workloads) I vote to take it into account before.
+
+## Design Details
+
+### Workload API
+
+```go
+type PodSet struct {
+    // .......
+
+    // count is the number of pods for the spec.
+    // +kubebuilder:validation:Minimum=1
+    Count int32 `json:"count"`
+
+    // minimumCount the minimum caount acceoted for this
+    // PodSet in case of partial admission.
+    // If not provided, partial admission is not accepted
+    // for this PodSet
+    // +optional
+    MinimumCount *int32 `json:"minimumCount"`
+}
+
+```
+
+### Scheduler / Flavorassignment
+
+I case the current workload dose not in the current available quota and any of its PodSets allow partial admission and the cont did not reach the `MinimumCount`, decrease the cont in one PodSet and try again. The accepted number of pods in each PodSet are published in `workload.Status.Admission.PodSetAssignments[*].ResourceUsage.Count`
+
+### Jobframework
+
+```diff
+type GenericJob interface {
+    // ...
+
+	
+-    // RunWithNodeAffinity will inject the node affinity extracting from workload to job and unsuspend the job.
++    // RunWithPodSetsInfo will inject the node affinity and podsSet counts extracted from workload to the job and unsuspend the job.
+-    RunWithNodeAffinity(nodeSelectors []PodSetNodeSelector)
++    RunWithPodSetsInfo(nodeSelectors []PodSetNodeSelector, podSetCounts []int32)
+-    // RestoreNodeAffinity will restore the original node affinity of job.
++    // RestorePodSetsInfo will restore the original node affinity of job.
+-    RestoreNodeAffinity(nodeSelectors []PodSetNodeSelector)
++    RestorePodSetsInfo(nodeSelectors []PodSetNodeSelector, podSetCounts []int32)
+
+    // ...
+}
+
+```
+
+### batch/Job controller
```

Comment:
What happens with the job when admitted? do we update parallelism and completions?
Also, completions is immutable unless parallelism is set at the same time.

https://kubernetes.io/docs/concepts/workloads/controllers/job/#elastic-indexed-jobs

Also only possible for Indexed Jobs. So we need to validate that in the webhook.

Any notes for MPIJob?

Reply:
> What happens with the job when admitted?

The way I see it, since the batch user opts in  for partial admission , he/she is aware that the `parallelism` can change at runtime and populate  `completions` with the actual number of successful runs needed.

> Also only possible for Indexed Jobs. So we need to validate that in the webhook.

I think it should work for non indexed jobs as well.

> Any notes for MPIJob?

No, I think is better to target the core part of Kueue and batch/Job as a reference implementation.

Reply:
> The way I see it, since the batch user opts in for partial admission , he/she is aware that the `parallelism` can change at runtime and populate `completions` with the actual number of successful runs needed.

That reminds me that there are actually two scenarios:
- highly parallel jobs, where every index can run on its own. Then just modifying parallelism is good enough.
- jobs with pod-to-pod communication. Here all pods need to start together, then both parallelism and completions need to change.

Reply:
> No, I think is better to target the core part of Kueue and batch/Job as a reference implementation.

In a design doc, it's sometimes useful to see the bigger picture. We don't want to get stuck with an API that only works for one use case.

Implementation PRs can certainly be separate for Job and MPIJob.

Reply:
>In a design doc, it's sometimes useful to see the bigger picture. We don't want to get stuck with an API that only works for one use case.
>
>Implementation PRs can certainly be separate for Job and MPIJob.

I've added a brief section related to MPIJob.

Reply:
> 

Oh I missed this.

Jobs with pod-to-pod communication are jobs where all pods within the job communicate with each other. In this case, it doesn't make sense to have `parallelism!=completions`.

So this gives us two different use cases of partial admission:
- I accept lower level of parallelism, but keep number of completions.
- I accept lower level of parallelism and it should remain equal to number of completions.

Reply:
In this case , we should not enforce anything on the  `completion` field in the webhook, and the user should ensure that `parallelism` == `completions` for the first case leave the `completions` nil for the second.

Reply:
added, however  I cannot say I've fond this behavior described [here](https://kubernetes.io/docs/concepts/workloads/controllers/job/#parallel-jobs)

Reply:
it's number 3, where it says "do not specify completions"

Reply:
I meant the one that needs to keep `parallelism` and `completions` in sync.

Reply:
There is nothing that keeps them in sync. It's just one way that users can use the fields.

But somewhat related are these:
- https://kubernetes.io/docs/concepts/workloads/controllers/job/#elastic-indexed-jobs
- https://kubernetes.io/docs/tasks/job/job-with-pod-to-pod-communication/